crates/proto/src/dnssec/dnssec_dns_handle/mod.rs

```diff
@@ -854,26 +852,55 @@ where
         ));
     }
 
-    // otherwise we need to recursively discover the status of DS up the chain,
-    //   if we find a valid DS, then we're in a Bogus state,
-    //   if we get ProofError, our result is the same
+    Err(ProofError::new(
+        Proof::Bogus,
+        ProofErrorKind::DsRecordShouldExist { name: zone },
+    ))
+}
 
-    let parent = zone.base_name();
-    if zone == parent {
-        // zone is `.`. do not call `find_ds_records(.., parent, ..)` or that will lead to infinite
-        // recursion
-        return Err(ProofError::new(
-            Proof::Bogus,
-            ProofErrorKind::DsRecordShouldExist { name: zone },
-        ));
+/// Checks whether a DS RRset exists for the given name or an ancestor of it.
+async fn find_ds_records<H>(
+    handle: &DnssecDnsHandle<H>,
+    zone: Name,
+    options: DnsRequestOptions,
+) -> Result<(), ProofError>
+where
+    H: DnsHandle + Sync + Unpin,
+{
+    let result = fetch_ds_records(handle, zone.clone(), options).await;
```

Comment:
It looks like we could roll this instance into the loop below, in the style of a C `do { .. } while ..`, by just doing a `zone = zone.base_name()` at the bottom of the loop?

Reply:
The handling of the `Ok(_)` arm is different between the initial name and the parent names, so they can't be combined.